File: 16/c/main.c
#include <stdio.h>
#include <assert.h>
#include <limits.h>
#include <string.h>

#define N 1000

struct packet {
  int version;
  int type;
  long value;
  struct packet* sub;  // sub packet is a linked list
  struct packet* next; // linked list for siblings
};

struct packet packets[N];
int packets_size = 0;

struct packet* newPacket() {
  assert(packets_size < N && "Increase N");

  return &packets[packets_size++];
}

int dehex(char hex) {
  if (hex <= '9') return hex - '0';
  return hex - 'A' + 10;
}

char hex[4096] = {0};
char bin[4096 * 4] = {0};
char* bin_ptr = bin;

void binify() {
  for (size_t i = 0; i < strlen(hex); i++) {
    int j = i * 4;
    int num = dehex(hex[i]);

    bin[j + 0] = (num & 8) ? '1' : '0';
    bin[j + 1] = (num & 4) ? '1' : '0';
    bin[j + 2] = (num & 2) ? '1' : '0';
    bin[j + 3] = (num & 1) ? '1' : '0';
  }
}

long dec(int num) {
  long curr = 1;
  long ret = 0;

  for (int i = num - 1; i >= 0; i--) {
    if (bin_ptr[i] == '1') {
      ret += curr;
    }
    curr *= 2;
  }

  bin_ptr +=  num;

  return ret;
}

char* get(int num) {
  char* ret = bin_ptr;
  bin_ptr += num;
  return ret;
}

struct packet* packify() {
  struct packet* packet = newPacket();

  int version = dec(3);
  int type = dec(3);
  packet->version = version;
  packet->type = type;

  // literal value
  if (type == 4) {
    char buff[1024] = {0};
    int buff_size = 0;

    while (get(1)[0] == '1') {
      for (int i = 0; i < 4; i++) {
        buff[buff_size++] = *(bin_ptr++);
      }
    }

    for (int i = 0; i < 4; i++) {
      buff[buff_size++] = *(bin_ptr++);
    }

    // last one
    char* temp = bin_ptr;
    bin_ptr = buff;
    long literal = dec(buff_size);
    bin_ptr = temp;

    packet->value = literal;
    return packet;
  }

  if (get(1)[0] == '0') {
    long size = dec(15);
    char* end = bin_ptr + size;

    packet->sub = packify();

    struct packet* curr = packet->sub;
    while (bin_ptr != end) {
      curr->next = packify();
      curr = curr->next;
    }
  } else {
    long size = dec(11);

    packet->sub = packify();
    struct packet* curr = packet->sub;
    for (int i = 0; i < size - 1; i++) {
      curr->next = packify();
      curr = curr->next;
    }
  }

  return packet;
}

int sumVersion(struct packet* curr) {
  if (!curr) return 0;

  int ret = 0;

  while (curr) {
    ret += curr->version;
    ret += sumVersion(curr->sub);

    curr = curr->next;
  }

  return ret;
}

long operate(struct packet* packet) {
  struct packet* curr = packet->sub;
  int type = packet->type;

  if (type == 0) {
    long ret = 0;
    while (curr) {
      ret += operate(curr);
      curr = curr->next;
    }
    return ret;
  }

  if (type == 1) {
    long ret = 1;
    while (curr) {
      ret *= operate(curr);
      curr = curr->next;
    }
    return ret;
  }

  if (type == 2) {
    long min = LLONG_MAX;
    struct packet* curr = packet->sub;

    while (curr) {
      long val = operate(curr);
      if (val < min) min = val;
      curr = curr->next;
    }

    return min;
  }

  if (type == 3) {
    long max = 0;
    struct packet* curr = packet->sub;

    while (curr) {
      long val = operate(curr);
      if (val > max) max = val;
      curr = curr->next;
    }

    return max;
  }

  if (type == 4) {
    return packet->value;
  }

  if (type == 5) {
    long l = operate(curr);
    long r = operate(curr->next);

    return l > r ? 1 : 0;
  }

  if (type == 6) {
    long l = operate(curr);
    long r = operate(curr->next);

    return l < r ? 1 : 0;
  }

  if (type == 7) {
    long l = operate(curr);
    long r = operate(curr->next);

    return l == r ? 1 : 0;
  }

  assert(0 && "Unreachable");
}

int main(int argc, char *argv[]) {
  assert(argc == 2 && "Must provide one argument => input.txt path");

  printf("Input file: %s\n\n", argv[1]);
  FILE *file = fopen(argv[1], "r");

  assert(fscanf(file, "%s", hex));


  binify();
  struct packet* genesis = packify();

  // Problem 1
  {
    printf("Answer 1: %d\n", sumVersion(genesis));
  }


  // Problem 2
  {
    printf("Answer 2: %ld\n", operate(genesis));
  }


  fclose(file);

  return 0;
}

File: 16/js/main.js
const fs = require('fs')
const {join} = require('path')

const file = fs.readFileSync(join(__dirname, '../input.txt'), 'utf8')
const input = file.split('').map(n => parseInt(n, 16).toString(2).padStart(4, '0')).join('')

function dec(n) {
  return parseInt(n, 2)
}


function getStructure(input) {
  function get(num) {
    const ret = input.input.slice(0, num)
    input.input = input.input.slice(num)
    return ret
  }

  const version = dec(get(3))
  const type = dec(get(3))

  // literal value
  if (type == 4) {
    let literal = ''
    while (get(1) == '1') {
      literal += get(4)
    }

    // last one
    literal += get(4)
    return {version, type: 'literal', value: dec(literal)}
  }

  const ret = {version, type: 'operator', operation: type, packets: []}
  if (get(1) == '0') {
    const len = dec(get(15))
    const packets = {input: get(len)}

    while (packets.input.length) {
      const packet = getStructure(packets)
      ret.packets.push(packet)
    }
  } else {
    const len = dec(get(11))

    for (let i = 0; i < len; i++) {
      const packet = getStructure(input)
      ret.packets.push(packet)
    }
  }

  return ret
}

const packets = getStructure({input})

// Problem 1
{
  let answer = 0
  function count({type, version, packets}) {
    answer += version

    if (type == 'operator') {
      for (packet of packets) {
        count(packet)
      }
    }
  }
  count(packets)

  console.log(`Answer 1: ${answer}`);
}

// Problem 2
{
  function operate({type, operation, packets, value}) {
    if (type == 'literal') return value

    if (type == 'operator') {
      if (operation == 0) { // sum
        let ret = 0
        for (packet of packets) {
          ret += operate(packet)
        }
        return ret
      }
      if (operation == 1) { // product
        let ret = 1
        for (packet of packets) {
          ret *= operate(packet)
        }
        return ret
      }
      if (operation == 2) { // min
        return Math.min(...packets.map(packet => operate(packet)))
      }
      if (operation == 3) { // max
        return Math.max(...packets.map(packet => operate(packet)))
      }
      if (operation == 5) { // less
        const l = operate(packets[0])
        const r = operate(packets[1])
        return l > r ? 1 : 0
      }
      if (operation == 6) { // greter
        const l = operate(packets[0])
        const r = operate(packets[1])
        return l < r ? 1 : 0
      }
      if (operation == 7) { // equal
        const l = operate(packets[0])
        const r = operate(packets[1])
        return l == r ? 1 : 0
      }
    }
  }

  console.log(`Answer 2: ${operate(packets)}`);
}
